src/Core/Domain/ValueObject/Reduction.php

```diff
@@ -0,0 +1,160 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\ValueObject;
+
+use PrestaShop\Decimal\Number;
+use PrestaShop\PrestaShop\Core\Domain\Exception\DomainConstraintException;
+
+/**
+ * Provides valid reduction values
+ */
+class Reduction
+{
+    /**
+     * For reducing concrete amount of money from price
+     */
+    const TYPE_AMOUNT = 'amount';
+
+    /**
+     * For reducing certain percentage calculated from price
+     */
+    const TYPE_PERCENTAGE = 'percentage';
+
+    /**
+     * Allowed reduction types
+     */
+    const ALLOWED_TYPES = [
+        self::TYPE_AMOUNT,
+        self::TYPE_PERCENTAGE,
+    ];
+
+    /**
+     * Maximum allowed value for percentage type reduction
+     */
+    const MAX_ALLOWED_PERCENTAGE = 100;
+
+    /**
+     * @var string
+     */
+    private $type;
+
+    /**
+     * @var Number
+     */
+    private $value;
+
+    /**
+     * @param string $type
+     * @param float $value
+     *
+     * @throws DomainConstraintException
+     */
+    public function __construct(string $type, float $value)
```

Comment:
I'm wondering since we use a number in the value object we should it use as the expected type in the constructor If not let's say we already have a string or even a Number as input it forces us to convert the value back to a float in order to reconstruct a Number which removes the interest of this class

Reply:
@eternoendless required that we only use native php primitive structures to make these as easy as possible to instantiate (but we must validate the data). We can add factory methods such as `createFromX()` to provide alternative constructors.